common/link.go

```diff
@@ -0,0 +1,64 @@
+package common
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
```

Comment:
Use a `NewLink(url string) (*Link, error)` function here instead of requiring other bits of code to manually build the struct and call `ValidateLink()` and `DetermineLinkType()` directly.